### PR TITLE
feat: support dark theme

### PR DESCRIPTION
**Pull request by agatha197**
resolves https://github.com/lablup/backend.ai-webui/issues/577
related Issue: https://github.com/lablup/backend.ai-webui/issues/2160

## What's changed
- This PR supports the dark theme.
  - You can toggle the theme by clicking the top bar icon.
	(dark -> light mode)
	image
	(light -> dark)
	image
  - The antd token values are saved to the react root page, such as `--general-<token name>: <token value>`. ([codes](https://github.com/lablup/backend.ai-webui/pull/2193/files#diff-60beaf4458ac35704896a1753028ec75516f554b79a92cd6ecaf679bec6e8ee9R68-R74))
    - And these values are used in web elements.
    - Remove most of the hardcoded color. And they are replaced with `--general-<token name>`.
    - For better readability, I set the primary color to `#37B076` in light mode, and `#25C2A0` in dark theme mode.
      -  you can check color code in `/resources/theme.json`
- Added react usersettings draft page. ([codes](https://github.com/lablup/backend.ai-webui/pull/2193/files#diff-80eae526b6641a005d21da1d22fb620b15d205d0c0742b8c8a21d8b41ac8dd6fR68-R119))
  - These codes will be replaced after all of usersetting features are implemented.
  - So far, the bootstrap script, edit user config script, etc are not yet implemented.

## Checklist
- [ ] Is there a case where you can't see the content because the color changes?
- [ ] Is the theme changed if you click the top bar toggle button?

**Checklist:** (if applicable)

- [ ] Mention to the original issue
- [ ] Documentation
- [x] Minium required manager version: 23.09
- [x] Specific setting for review (eg., KB link, endpoint or how to setup)
- [x] Minimum requirements to check during review
- [ ] Test case(s) to demonstrate the difference of before/after

### CODE REVIEW

**Comment by yomybaby** [comment]
image
We can make the scrollbar also affected by dark mode.  (ref: https://react.dev/)

**Comment by yomybaby** [comment]
image
It is a selected menu item, but it stands out too much in dark mode. How about making some adjustments?

**Comment by agatha197** [comment]
> https://github.com/lablup/backend.ai-webui/pull/2193#issuecomment-1961072232
> https://github.com/lablup/backend.ai-webui/pull/2193#issuecomment-1961077922

All resolved ;)

File: react/src/components/MainLayout/MainLayout.tsx [hunk 1]
```diff
@@ -75,6 +75,31 @@ function MainLayout() {
...
+
+          *::-webkit-scrollbar-track {
+            background: var(--general-colorBgContainer, transparent);
+          }
+
+          *::-webkit-scrollbar-thumb {
+            background-color: var(--general-colorBorderSecondary, #464646);
+          }
+        `}
+      </style>
```
Comment (yomybaby) [inline react/src/components/MainLayout/MainLayout.tsx]: Thank you. It's better than before.  But it loose background color or scrollbar. 
image
image

Please change it like this.
image
image

File: resources/theme.json [hunk 2]
```diff
@@ -16,12 +16,15 @@
...
+      },
+      "Layout": {
+        "lightSiderBg": "#F4F7FD"
       }
     }
   },
   "dark": {
     "token": {
-      "colorPrimary": "#37B076"
+      "colorPrimary": "#25C2A0"
```
Comment (yomybaby) [inline resources/theme.json]:
```suggestion
     "fontFamily": "'Ubuntu', Roboto, sans-serif",
      "colorPrimary": "#25C2A0"
```
Don't forget to set the same `fontFamily` style as the light mode. Different font families may cause subtle layout changes on the screen.

**Review by yomybaby — CHANGES_REQUESTED** [review]
Let's match the background color of the progress bar for dark mode between Web Component and React component.
image

Example:
image

File: resources/theme.json (hunk 2 above)
Comment (yomybaby) [reply resources/theme.json]: If the fontFamily is not matched, the dark mode toggle button moves slightly when clicked.

**Review by yomybaby — CHANGES_REQUESTED** [review]
| Web component | React |
|----|---|
| image  | image | 

In dark mode, the background color of the Web component modal is not distinguished from the mask area.

The React modal is better distinguished because it has:
- A lighter background for the modal
- A darker background for the modal mask area

**Comment by yomybaby** [comment]
@agatha197 I pushed a commit to introduce a theme configuration for enforcing the sidebar theme mode. Now, BAISider works with the theme prop. It no longer needs to determine whether dark mode is being used using useThemeMode.
If you want to enforce theme mode of sider, set the `sider` option to`theme.json`
```
{
  "light": {},
  "dark": {},
  "sider": {
    "theme":"dark"
  }
}
```

File: react/src/components/Flex.tsx [hunk 3]
```diff
@@ -65,7 +65,6 @@ const Flex = React.forwardRef<HTMLDivElement, FlexProps>(
         ref={ref}
         style={{
           alignItems: 'stretch',
-          backgroundColor: 'transparent',
```
Comment (yomybaby) [inline react/src/components/Flex.tsx:68]: After deleting this line, we need to update the snapshots of Flex.test.tsx.
You can refer to the [documentation](https://jestjs.io/docs/snapshot-testing#interactive-snapshot-mode) on how to update snapshots.

**Review by yomybaby — CHANGES_REQUESTED** [review]

File: react/package-lock.json [hunk 4]
```diff
@@ -136,12 +136,12 @@
       }
     },
     "node_modules/@ant-design/icons": {
-      "version": "5.2.6",
-      "resolved": "https://registry.npmjs.org/@ant-design/icons/-/icons-5.2.6.tgz",
-      "integrity": "sha512-4wn0WShF43TrggskBJPRqCD0fcHbzTYjnaoskdiJrVHg86yxoZ8ZUqsXvyn4WUqehRiFKnaclOhqk9w4Ui2KVw==",
+      "version": "5.3.0",
```
Comment (yomybaby) [inline react/package-lock.json:139]: Only `package-lock.json` has been updated. If you want to upgrade `@ant-design/icons`, please make sure to update `package.json` as well.

**Review by yomybaby — CHANGES_REQUESTED** [review]

File: react/src/components/DefaultProviders.tsx [hunk 5]
```diff
@@ -118,7 +119,7 @@ export const useCurrentLanguage = () => {
     return () => window.removeEventListener('langChanged', handler);
   }, [i18n]);
 
-  return [lang] as const;
+  return { lang, setLang } as const;
```
Comment (yomybaby) [inline react/src/components/DefaultProviders.tsx]: Is there a reason to change this part from an array to an object?
`[lang, setLang]` vs. `{lang, setLang}`

File: react/src/components/UserSettings.tsx [hunk 6]
```diff
@@ -0,0 +1,310 @@
+import { useLocalStorageGlobalState } from '../hooks/useLocalStorageGlobalState';
```
Comment (yomybaby) [inline react/src/components/UserSettings.tsx]: The `UserSetting` page is not completed yet, so it is not being used in this PR, right?

File: react/src/components/MainLayout/MainLayout.tsx [hunk 7]
```diff
@@ -64,12 +65,41 @@ function MainLayout() {
     };
   }, [navigate]);
 
+  // Add antd token styles to react root for web components
+  _.forEach(token, (value, key) => {
+    document.documentElement.style.setProperty(
+      `--general-${key}`,
```
Comment (yomybaby) [inline react/src/components/MainLayout/MainLayout.tsx]: `--general-` is a namespace used in the web component part. I am concerned about duplicated names. How about renaming it? `--token-`?

File: react/src/components/ModelCardModal.tsx [hunk 8]
```diff
@@ -299,8 +299,16 @@ const ModelCardModal: React.FC<ModelCardModalProps> = ({
...
+            // styles={{
+            //   body: {
+            //     padding: token.paddingLG,
+            //     overflow: 'auto',
+            //     height: screen.lg ? 'calc(100vh - 243px)' : undefined,
+            //     minHeight: 200,
+            //     // maxHeight: 650
+            //   },
+            // }}
           >
```
Comment (yomybaby) [inline react/src/components/ModelCardModal.tsx]: This code is not related to 'dark mode'. Let's remove (revert) this part or add a `//TODO:` commnet

File: react/src/components/SSHKeypairGenerationModal.tsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
+import { useSuspendedBackendaiClient } from '../hooks';
```
Comment (yomybaby) [inline react/src/components/SSHKeypairGenerationModal.tsx:1]: Can you remove the files related to the user setting page that you created and create a separate branch?

File: react/src/pages/SessionListPage.tsx [hunk 10]
```diff
@@ -106,7 +106,7 @@ const SessionListPage: React.FC<PropsWithChildren> = ({ children }) => {
             <p>SessionList</p>
           </Card> */}
 
-          {/* <Card bodyStyle={{ paddingTop: 0 }}> */}
+          {/* <Card bodyStyles={{paddingTop: 0}}> */}
```
Comment (yomybaby) [inline react/src/pages/SessionListPage.tsx]: This change is not necessary for this PR, please revert it.

File: react/src/components/WebUIThemeToggleButton.tsx [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  return (
+    <Button
+      size="large"
+      type="text"
+      icon={isDarkMode ? <SunOutlined /> : <MoonOutlined />}
+      onClick={() => {
+        const event: Event = new CustomEvent('theme-mode-changed', {
+          detail: { token },
+        });
+        document.dispatchEvent(event);
```
Comment (yomybaby) [inline react/src/components/WebUIThemeToggleButton.tsx]: It would be better to move this part, which dispatches an event when `setThemeMode`, to `hooks/useThemeMode.tsx` for better separation of concerns.

**Comment by gee05053** [comment]
URL is not visible well in dark mode.

Light mode
스크린샷 2024-02-26 오전 11 16 23


Dark mode
스크린샷 2024-
02-26 오전 11 12 01

**Review by gee05053 — CHANGES_REQUESTED** [review]
URL is not visible well in dark mode.

Light mode
스크린샷 2024-02-26 오전 11 16 23
스크린샷 2024-02-26 오전 11 28 54



Dark mode
스크린샷 2024-
02-26 오전 11 12 01
스크린샷 2024-02-26 오전 11 07 26

**Review by gee05053 — CHANGES_REQUESTED** [review]
How about unifing button's color in dark mode

problem
스크린샷 2024-02-26 오전 11 05 13

like this
스크린샷 2024-02-26 오전 11 24 33

**Review by gee05053 — CHANGES_REQUESTED** [review]
React components is not applied language.
스크린샷 2024-02-26 오전 11 10 07

**Review by gee05053 — CHANGES_REQUESTED** [review]
How about unifing icon's color in dark mode at login dialog.
Some icon is colored and Some is black and white.

스크린샷 2024-02-26 오전 11 06 28

File: src/components/backend-ai-splash.ts [hunk 12]
```diff
@@ -114,7 +113,13 @@ export default class BackendAISplash extends LitElement {
...
     return html`
       <backend-ai-dialog id="splash-panel" fixed backdrop blockscrolling persistent narrowLayout hideActions>
-        <div class="splash-header" slot="title">
+        <div class="splash-header" slot="title"
+          style="background-image:url(${
+            localStorage.getItem('backendaiwebui.settings.isDarkMode')
+              ? 'manifest/backend.ai-text-bgdark.svg'
+              : 'manifest/backend.ai-text.svg'
+          });"
+        >
```
Comment (yomybaby) [inline src/components/backend-ai-splash.ts:146]: Please check `About` modal text:
image

Please change the color of `Release Note License` and `open-source software` for dark mode.

File: src/components/backend-ai-webui.ts [hunk 13]
```diff
@@ -1895,7 +1898,7 @@ export default class BackendAIWebUI extends connect(store)(LitElement) {
             </div>
             <address class="full-menu">
               <small class="sidebar-footer">Lablup Inc.</small>
-              <small class="sidebar-footer" style="font-size:9px;">24.03.0-alpha.4.5862</small>
+              <small class="sidebar-footer" style="font-size:9px;">24.03.0-alpha.4.5927</small>
```
Comment (yomybaby) [inline src/components/backend-ai-webui.ts]: Please revert this line.
```suggestion
              <small class="sidebar-footer" style="font-size:9px;">24.03.0-alpha.4.5862</small>
```

**Review by yomybaby — CHANGES_REQUESTED** [review]

**Comment by agatha197** [comment]
> How about unifing icon's color in dark mode at login dialog. Some icon is colored and Some is black and white.
> 
> 스크린샷 2024-02-26 오전 11 06 28

That's intended. Because left buttons are just icon, and right buttons are clickable buttons.

**Comment by agatha197** [comment]
> React components is not applied language. 스크린샷 2024-02-26 오전 11 10 07

It would be better to handle that issue in another PR. Please make a issue for that :)

**Review by gee05053 — CHANGES_REQUESTED** [review]
If user toggle light and dark mode, color of componentes is remained  dark mode.

main branch
스크린샷 2024-02-26 오후 1 05 38

this branch after toggling theme mode
스크린샷 2024-02-26 오후 1 06 46
스크린샷 2024-02-26 오전 11 09 50
스크린샷 2024-02-26 오후 1 06 49

**Review by gee05053 — CHANGES_REQUESTED** [review]
The background of the select dropdown on the statistic page does not change when in dark mode.
스크린샷 2024-02-26 오후 1 13 34

**Review by gee05053 — CHANGES_REQUESTED** [review]
In the Add Policy dialog on the resource policy tab on the credential page, select has a box-shadow.
스크린샷 2024-02-26 오후 1 19 01

**Review by gee05053 — CHANGES_REQUESTED** [review]
<b>Environment Page</b>
I think It is better there was a distinction between input that is impossible to input and input that can be input but has not been input yet.

Problem
스크린샷 2024-02-26 오후 1 27 06
스크린샷 2024-02-26 오후 1 27 19
스크린샷 2024-02-26 오후 1 27 41

Like
스크린샷 2024-02-26 오후 1 28 09

**Review by gee05053 — CHANGES_REQUESTED** [review]
<b>Resources Page</b>

1. Dialog has a line like scroll but It is not scroll.
스크린샷 2024-02-26 오후 1 43 42
스크린샷 2024-02-26 오후 1 43 48

2. select's background color is different.
스크린샷 2024-02-26 오후 1 44 05

**Review by gee05053 — CHANGES_REQUESTED** [review]
<b>Configurations Page</b>

How about changing the color of the disabled switch to make it more distinguishable?
스크린샷 2024-02-26 오후 1 48 00

**Review by gee05053 — CHANGES_REQUESTED** [review]
<b>Session Page</b>

Please check icon's color
스크린샷 2024-02-26 오후 2 07 17
스크린샷 2024-02-26 오후 2 10 51

**Review by gee05053 — CHANGES_REQUESTED** [review]
<b>Data&Storage Page</b>

Some text is not applied dark mode.

스크린샷 2024-02-26 오후 2 17 39
스크린샷 2024-02-26 오후 2 18 00
스크린샷 2024-02-26 오후 2 18 17
스크린샷 2024-02-26 오후 2 18 27

**Review by gee05053 — CHANGES_REQUESTED** [review]
I suggest that when login dialog is open, background page is also applied dark mode.

스크린샷 2024-02-26 오후 2 22 25

File: react/src/components/DefaultProviders.tsx [hunk 14]
```diff
@@ -90,9 +90,10 @@ i18n
   });
 
 export const useCurrentLanguage = () => {
-  const [lang, _setLang] = useState(
+  const [lang, setLang] = useState(
     //@ts-ignore
-    globalThis?.backendaioptions?.get('current_language'),
+    // globalThis?.backendaioptions?.get('current_language'),
+    localStorage.getItem('current_language') || 'en',
```
Comment (gee05053) [inline react/src/components/DefaultProviders.tsx]: @agatha197  Please check this code to change language option.

[Comment](https://github.com/lablup/backend.ai-webui/pull/2193#pullrequestreview-1899865850) that I wrote is related to this line.

File: react/src/components/DefaultProviders.tsx (hunk 5 above)
Comment (agatha197) [reply react/src/components/DefaultProviders.tsx]: no... I'll revert that.

File: react/src/components/SSHKeypairGenerationModal.tsx (hunk 9 above)
Comment (agatha197) [reply react/src/components/SSHKeypairGenerationModal.tsx:1]: okay. I'll make a new PR :)

File: react/src/components/UserSettings.tsx (hunk 6 above)
Comment (agatha197) [reply react/src/components/UserSettings.tsx]: yes. I'll handle this in another PR.

**Comment by agatha197** [comment]
> I suggest that when login dialog is open, background page is also applied dark mode.
> 
> 스크린샷 2024-02-26 오후 2 22 25

I cannot see the image

File: react/src/components/BAISider.tsx [hunk 15]
```diff
@@ -66,79 +67,85 @@ const BAISider: React.FC<BAISiderProps> = ({
           left: 0,
           borderRight: '1px solid',
           borderColor: token.colorBorder,
-          backgroundColor: '#F4F7FD',
           paddingTop: token.paddingContentVerticalSM,
```
Comment (gee05053) [inline react/src/components/BAISider.tsx:70]:
```suggestion
          paddingTop: token.paddingContentVerticalSM,
          scrollbarColor: 'auto'
```

If you want to delete scrollbar at sider, add this line.

**Review by gee05053 — CHANGES_REQUESTED** [review]
I found the icon color issue
Please check this.
스크린샷 2024-02-26 오후 10 42 13

**Review by gee05053 — CHANGES_REQUESTED** [review]
I found the issue at serveral page that is not applied dark mode
You can test the issue in the following way.

1. Login and set dark mode.
2. Set light mode again.
3. Refresh browser using `command + R`
4. Set dark mode again
5. Visit Session, Data&Storage, Statistics, Credential ... page

스크린샷 2024-02-26 오후 10 46 08
스크린샷 2024-02-26 오후 10 46 16
스크린샷 2024-02-26 오후 10 46 19
스크린샷 2024-02-26 오후 10 46 25
스크린샷 2024-02-26 오후 10 46 28
스크린샷 2024-02-26 오후 10 46 31
스크린샷 2024-02-27 오후 1 03 32
스크린샷 2024-02-27 오후 1 42 48

**Comment by agatha197** [comment]
> I found the icon color issue Please check this. 스크린샷 2024-02-26 오후 10 42 13

It's image. not an icon. I think it's better to leave it as it is.

**Review by yomybaby — CHANGES_REQUESTED** [review]
After this PR, the web component cards have more shadow. Could you revert that style?
| Before | After |
|--------|--------|
| image | image  |

**Review by yomybaby — CHANGES_REQUESTED** [review]
Please check the layout of Download card on `/summary`.  ( It's okay that the button color has changed. )
| Before | After |
|--------|--------|
| image | image |

**Comment by yomybaby** [comment]
> After this PR, the web component cards have more shadow. Could you revert that style?
> 
> Before	After
> image	image

We don't need to make it the same as before. It's better to remove the shadow to match the React Card component style. @agatha197

**Review by gee05053 — CHANGES_REQUESTED** [review]
Environment Page

Is the button text color intentional?
I found that mwc-dialog in Environment page and userSetting page
스크린샷 2024-02-27 오전 9 59 22
스크린샷 2024-02-27 오전 9 59 29

**Comment by agatha197** [comment]
> After this PR, the web component cards have more shadow. Could you revert that style?
> 
> Before	After
> image	image

I'll remove the box shadow like antd component!
image
image

**Review by gee05053 — CHANGES_REQUESTED** [review]
`About Backen.AI` dialog's logo is appeared dark mode design at light mode.
스크린샷 2024-02-27 오후 1 47 51

You can test the issue in the following way.
1. Set dark mode and open `About backend.AI` dialog.
2. Close dialog and Refresh page using `cmd + R`
3. open `About backend.AI` dialog again.

**Review by gee05053 — CHANGES_REQUESTED** [review]
<b>Only Electron Issue</b>

1. Lit element's scrollbar-track is not applied dark mode.
스크린샷 2024-02-27 오후 1 45 10

2. Textfield's backgroundColor is different from dialog backgroundColor
스크린샷 2024-02-27 오후 1 45 28

**Review by gee05053 — CHANGES_REQUESTED** [review]
message is not applied dark mode.

UserProfileSettingModal in UserDropdownMenu
스크린샷 2024-02-27 오후 1 56 25

UserSettingModal in Credential page
스크린샷 2024-02-27 오후 1 57 03

**Review by gee05053 — CHANGES_REQUESTED** [review]
Cancelled Session info text is not applied dark mode.

스크린샷 2024-02-27 오후 3 10 10

**Review by gee05053 — CHANGES_REQUESTED** [review]
Check this Please🙏🏻

스크린샷 2024-02-27 오후 8 36 40

**Comment by agatha197** [comment]
> message is not applied dark mode.
> 
> UserProfileSettingModal in UserDropdownMenu 스크린샷 2024-02-27 오후 1 56 25
> 
> UserSettingModal in Credential page 스크린샷 2024-02-27 오후 1 57 03

This issue will be automatically resolved after implementing react usersetting page.

**Review by yomybaby — APPROVED** [review]
LGTM! 💪💪
